### PR TITLE
Fix var highlighting

### LINKED ISSUE #85: Parameterized types do not highlight variable name
`List list` works fine, correctly highlighting list, however `List<Map> list` doesn't highlight list at all.

screenshot_20170809_202712

### PR DESCRIPTION
**Pull request by vincekd**
highlight variable declarations and assignments

[fix #85]

### CODE REVIEW

**Comment by vincekd** [comment]
so I'm not sure if this is the best way to do this, but I folded all variable highlighting into a single function that works correctly even in some edge cases:
* `def (a, b, c) = list`
* `List<Map<String, Object>> foo`
* `if ((s = getS()) {`
* `[:].each { String key, def val ->`


Also added a bunch of unit tests for highlighting variables.

File: groovy-mode.el [hunk 1]
```diff
@@ -329,7 +312,77 @@ The function name is the second group in the regexp.")
...
+                (equal (match-string 3) "/*"))
+        (setq matched (- matched 2)))
+      (or (and (not (groovy--in-string-p))
+               (not (groovy--comment-p matched))
+               (let* ((line-beg (- (match-beginning 0) (line-beginning-position)))
+                      (cur-line (groovy--current-line))
+                      (str (buffer-substring-no-properties (line-beginning-position) (match-beginning 0)))
+                      (match-str (s-trim (match-string 0)))
+                      (open-char (match-string 1)))
+                 (message "match-str: %s" match-str)
```
Comment (Wilfred) [inline groovy-mode.el]: Debugging code I presume.

File: test/unit-test.el [hunk 2]
```diff
@@ -172,6 +172,7 @@ then run BODY."
    (should (eq (face-at-point) 'font-lock-keyword-face))))
 
 (defun faces-at-point ()
+  "Return face at point."
```
Comment (Wilfred) [inline test/unit-test.el]: Maybe say face*s* here, as this returns all faces at the current position.

File: test/unit-test.el [hunk 3]
```diff
@@ -229,3 +230,78 @@ then run BODY."
    "def bar /* foo */"
    (search-forward "foo")
    (should (not (memq 'font-lock-string-face (faces-at-point))))))
+
+(ert-deftest groovy-highlight-variable-assignment ()
```
Comment (Wilfred) [inline test/unit-test.el]: I don't think we should highlight assignment to existing varaibles: this face should only be for declaring new variables. This is consistent with elisp highlighting:

```emacs-lisp
;; foo is highlighted
(defvar foo 1)

;; foo is not highlighted
(setq foo 1)
```

File: groovy-mode.el [hunk 4]
```diff
@@ -329,7 +312,77 @@ The function name is the second group in the regexp.")
...
+                            (let ((start
+                                   (if (equal open-char "(")
+                                       line-beg
+                                     (when (string-match (rx "(" (+ (not (any ")"))) eol) str)
+                                       (match-end 0)))))
+                              (or (not start)
+                                  (string-match (rx bol (+ (not (any ")"))) ")" (* space) (or "=" "{"))
+                                                (substring cur-line start nil)))))))))))
+          ;; keep searching
+          (groovy--variable-names-search limit)))))
```
Comment (Wilfred) [inline groovy-mode.el]: This function is calling itself recursively, and terminating when we reach the limit or find a variable. However, Emacs doesn't have TCO, so this risks stack overflow on large files with lots of variables. 

Could we use a while loop instead? We do this elsewhere: https://github.com/Groovy-Emacs-Modes/groovy-emacs-modes/blob/606d6216c291c96f948e04ca481962c326771cf7/groovy-mode.el#L268-L270

File: groovy-mode.el [hunk 5]
```diff
@@ -329,7 +312,77 @@ The function name is the second group in the regexp.")
...
+      (or (and (not (groovy--in-string-p))
+               (not (groovy--comment-p matched))
+               (let* ((line-beg (- (match-beginning 0) (line-beginning-position)))
+                      (cur-line (groovy--current-line))
+                      (str (buffer-substring-no-properties (line-beginning-position) (match-beginning 0)))
+                      (match-str (s-trim (match-string 0)))
+                      (open-char (match-string 1)))
+                 (message "match-str: %s" match-str)
+                 (save-match-data
+                   (unless (string-match (rx bol (* space) (eval match-str) (* space) eol) cur-line)
```
Comment (Wilfred) [inline groovy-mode.el]: This isn't quite right: `rx` is evaluated at compile time, and match-str isn't defined at that point. I think you want to use `rx-to-string` with backquoting.

File: groovy-mode.el (hunk 4 above)
Comment (vincekd) [inline groovy-mode.el]: Heya! Thanks for the feedback!

The keyword funcs are passed a limit that (from what I've seen) is just outside the buffer's viewable area, so usually until you scroll no more than what can fit on the page is parsed and highlighted. I looked at some other modes and this was the most common way to implement it.

File: test/unit-test.el (hunk 3 above)
Comment (vincekd) [inline test/unit-test.el]: Do you think it would be worthwhile to set a defcustom option for this? Some people (me) might like it.

Comment (Wilfred) [inline test/unit-test.el]: Emacs is all about accommodating different people's preferences :)

A defcustom sounds like a good idea. Something like `groovy-highlight-assignments`?
